examples/remote/client.rs

```diff
@@ -0,0 +1,59 @@
+//! A simple command line client that allows issuing queries to a remote Bevy
+//! app via the BRP.
+
+use anyhow::Result as AnyhowResult;
+use argh::FromArgs;
+use bevy::remote::{DEFAULT_ADDR, DEFAULT_PORT};
+
+/// Struct containing the command-line arguments that can be passed to this example.
+/// The components are passed by their full type names positionally, while `host`
+/// and `port` are optional arguments which should correspond to those used on
+/// the server.
+///
+/// When running this example in conjunction with the `server` example, the `host`
+/// and `port` can be left as their defaults.
+///
+/// For example, to connect to port 1337 on the default IP address and query for entities
+/// with `Transform` components:
+/// ```text
+/// cargo run --example client -- --port 1337 bevy_transform::components::transform::Transform
+/// ```
+#[derive(FromArgs)]
+struct Args {
+    /// the host IP address to connect to
+    #[argh(option, default = "DEFAULT_ADDR.to_string()")]
+    host: String,
+    /// the port to connect to
+    #[argh(option, default = "DEFAULT_PORT")]
+    port: u16,
+    /// the full type names of the components to query for
+    #[argh(positional, greedy)]
+    components: Vec<String>,
+}
+
+/// The application entry point.
+fn main() -> AnyhowResult<()> {
+    // Parse the arguments.
+    let args: Args = argh::from_env();
+
+    // Create the URL. We're going to need it to issue the HTTP request.
+    let host_part = format!("{}:{}", args.host, args.port);
+    let url = format!("http://{}/", host_part);
+
+    let res = ureq::post(&url)
+        .send_json(ureq::json!({
```

Comment:
Is this meant to show users specifically that "it's just JSON, so any language would work, it just happens to be Rust here"?

Because if not then I don't see why this does not leverage the type info of the request.
I see that `.send_json` takes `impl Serialize`, so passing a `BrpRequest` here would work.

Looking it up I see that the fields within `BrpRequest` are just `serde_json::Value` types right now so constructing a req is a bit awkward.
Maybe clients aren't expected to have any type info to lean on here, curious what others think?

Reply:
An assortment of opinions for you:

1. Clients can be written in any language.
2. Rust clients should be assumed to be aware of the `brp` types, but not any component / resource types.
3. We should demonstrate how to best do this in Rust code in our examples.

Reply:
At least as currently designed, `BrpRequest` is not really the ideal thing for the client side, since it delays deserializing the fields for the sake of error reporting (and doesn't restructure them back into a "finished thing" before processing the result of that transformation). Honestly, I'm not sure exactly how valuable or load-bearing that choice is; it's probably worth tinkering with.

On the other hand, the types like `BrpGetParams` in `builtin_methods` could plausibly be used to make constructing the parameters for the requests more straightforward, although they were not really designed with that in mind. (In general, the Rust types are intended primarily for use on the server side.)

Reply:
Okay, I tinkered around with this and reached the conclusion that there isn't actually any reason to defer parsing on the `method` field, but the other fields are still basically required to be arbitrary optional JSON values in order to support the JSON-RPC spec. With some tweaks to the serialization derive, that should also be roughly suitable for client-side use (at least in the `client` example).

Reply:
Taking *another* another look at this, it seems that maybe the reason it is structured this way is so that the `id` for the request can be reported if it is successfully parsed, even if other fields are missing and/or incorrect. I think that's actually fairly valuable, so I'm going to stick with that philosophy for now. Maybe in the future, it would be possible to restructure these in a way that makes the server and client side data types line up nicely.

Reply:
When I changed this my only intention was to do something more reasonable than [manually constructing a string of json](https://github.com/bevyengine/bevy/pull/14880/files/ac96b055e8bb97b588d1ec221d6c29794b382e60#diff-b9891fe288692251ee4e78ed3d86ebbd13b187554d075ceb9f7e8d067d5fcce7R67-R76) so no strong attachment to this implementation from that perspective.